w3af/plugins/audit/open_api_auth.py

```diff
@@ -0,0 +1,372 @@
+"""
+open_api_auth.py
+
+Copyright 2018 Andres Riancho
+
+This file is part of w3af, http://w3af.org/ .
+
+w3af is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation version 2 of the License.
+
+w3af is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with w3af; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+"""
+import copy
+
+import w3af.core.data.kb.knowledge_base as kb
+import w3af.core.controllers.output_manager as om
+
+from w3af.core.controllers.plugins.audit_plugin import AuditPlugin
+from w3af.core.data.constants import severity
+from w3af.core.data.kb.vuln import Vuln
+
+
+class open_api_auth(AuditPlugin):
+    """
+    Checks if REST API endpoints authenticate incoming requests
+    as defined in their API specification.
+    See details in the plugin description below.
+
+    Here is a list of possible enhancements:
+      * Consider sending requests with other HTTP methods,
+        and check that 401 or 405 error code is returned
+        On the one hand, it may not look too good if we just replace the method,
+        and keep the rest of the request untouched (headers, payload, etc).
+        But on the other hand, the application is supposed to check HTTP method
+        in the very beginning, and reject requests if the method is not supported.
+        If the API spec says a particular method is supported,
+        then the check should expect 401
+        If the API spec says a particular method is not supported,
+        then the check should expect 405 (preferred) or 401
+        (should it only expect 405 in this case?).
+      * Allow a user to specify response codes that the plugin should expect.
+        Currently the plugin expects only 401.
+        We can introduce 'expected_code_regex' configuration parameter
+        which set a regex for expected response codes.
+      * If the plugin is updated to send other HTTP methods (see above),
+        we may want to introduce another configuration parameter
+        to set expected response codes in this case.
+      * Severity of identified issues may depend on response codes
+        returned by the server.
+
+    :author: Artem Smotrakov (artem.smotrakov@gmail.com)
+    :author: Andres Riancho (andres.riancho@gmail.com)
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self._spec = None
+        self._expected_codes = [401]
```

Comment:
Maybe add 403?

Reply:
I also thought about it. In a perfect world, an application should return 401 if no auth info is present at all. 403 should be usually returned if a request contains auth info, but authorization failed. So now the plugin assumes that it lives in a perfect world :) I think it is not a big deal If an application returns 403 instead of 401 but strictly speaking it doesn't sound correct to me. Anyway, I don't mind adding 403 here. I would even let a user to configure it - it may be next improvement.